sql: Add exclude_data_from_backup storage option to ALTER TABLE

adityamaru opened the pull request:
This change adds a storage parameter `exclude_data_from_backup`
to mark a table's row data as excluded from a backup.

`ALTER TABLE foo SET (exclude_data_from_backup = true);`

The above ALTER TABLE stmt sets the `ExcludeDataFromBackup` field
on a table descriptor

ExcludeDataFromBackup specifies if the table's row data can be excluded
from a backup targeting this table. This in turn means that the protected
timestamp record written during the backup will exclude this table as a
target, thereby not holding up GC on this table for the runtime of the
backup. This is particularly useful for table's with high-churn that can
then be configured with a low GC-TTL.

It is important to note that we will still backup the table descriptor, and
all other objects that reference this table egs: jobs, and so restoring
such a table will result in an empty table on the restoring cluster.

This change does not teach any part of the system to use this information
from the table descriptor, that will come in follow up PRs.

Informs: #73536

Release note (sql change): Adds new `ALTER TABLE <table> SET (exclude_data_from_backup = true);`
syntax. This allows a user to exclude row data for a particular table during a backup. On restoring this backup, the table will be created in the restoring cluster but will be empty. This flag also prevents a cluster or database backup from holding up the garbage collection time to live (GC TTL) on the key span representing this table. This is particularly useful for high churn tables that wish to keep a much shorter GC TTL to prevent the buildup of garbage.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75295)
<!-- Reviewable:end -->

adityamaru committed:
sql: Add `exclude_data_from_backup` storage option to ALTER TABLE

This change adds a storage parameter `exclude_data_from_backup`
to mark a table's row data as excluded from a backup.

`ALTER TABLE foo SET (exclude_data_from_backup = true);`

The above ALTER TABLE stmt sets the `ExcludeDataFromBackup` field
on a table descriptor

ExcludeDataFromBackup specifies if the table's row data can be excluded
from a backup targeting this table. This in turn means that the protected
timestamp record written during the backup will exclude this table as a
target, thereby not holding up GC on this table for the runtime of the
backup. This is particularly useful for table's with high-churn that can
then be configured with a low GC-TTL.

It is important to note that we will still backup the table descriptor, and
all other objects that reference this table egs: jobs, and so restoring
such a table will result in an empty table on the restoring cluster.

This change does not teach any part of the system to use this information
from the table descriptor, that will come in follow up PRs.

Informs: #73536

Release note: None

adityamaru commented:
@otan I'm going to hold off on merging this until your commits have landed so that I don't step on your toes. I just wanted some syntax in place so that I can build https://github.com/cockroachdb/cockroach/pull/75451 on top of this.

otan commented:
> @otan I'm going to hold off on merging this until your commits have landed so that I don't step on your toes

SGTM!

adityamaru commented on pkg/sql/paramparse/paramobserver.go:
```diff
@@ -144,6 +156,85 @@ var tableParams = map[string]tableParam{
+					"cannot set data in a temporary table to be excluded from backup")
+			}
+
+			// TODO(adityamaru): Do we need to check any privileges?
```
oh sweet, removed todo.

adityamaru commented on pkg/sql/logictest/testdata/logic_test/exclude_data_from_backup:
```diff
@@ -0,0 +1,68 @@
+ALTER TABLE t SET (exclude_data_from_backup = true);
+
+query T
+SELECT crdb_internal.pb_to_json('cockroach.sql.sqlbase.Descriptor', descriptor)->'table'->'excludeDataFromBackup'
```
nice, doing that now.

otan commented on pkg/sql/show_create.go:
```diff
@@ -181,6 +181,10 @@ func ShowCreateTable(
 	}
 
+	if exclude := desc.GetExcludeDataFromBackup(); exclude {
+		f.Buffer.WriteString(` WITH (exclude_data_from_backup = true)`)
```
can you grab https://github.com/cockroachdb/cockroach/commit/8d425f6c012072c82a4d3b43c95daa316956170c and make this line `storageParams = append(storageParams, `exclude_data_from_backup = true`)` instead? this will mean we don't double print WITH clauses for TTL + excluding data from backup

otan commented on pkg/sql/paramparse/paramobserver.go:
```diff
@@ -195,6 +195,41 @@ var tableParams = map[string]tableParam{
+				return errors.New("cannot set data in a table with inbound foreign key constraints to be excluded from backup")
+			}
+
+			s, err := GetSingleBool(key, datum)
```
do you want to allow `'on'` as well, as we do with outer variables?

if so, you may want to pull in https://github.com/cockroachdb/cockroach/commit/3bc72cfa9b2c46466f7a33086747ea84592271b3 and use `boolFromDatum`

adityamaru replied:
Hmm not sure  `exclude_data_from_backup = on` makes sense so leaving this as is for now

otan replied:
it also allows `'true'` as a string, and is standard for session vars (e.g. `set experimental_enable_temp_tables = 'on'`)

otan replied:
(although at the end of the day up to you)

adityamaru replied:
done! sprinkled 'on' and 'true' in the logictest too.

otan reviewed (approved):
minor suggestions with easy fixes; see #75779 if you need a ref to pull from.

adityamaru commented:
TFTR!

bors r=otan

adityamaru commented:
cc: @kathancox I just updated this Release Note now that the feature is complete and being shipped in 22.1.

kathancox commented:
Hey @adityamaru, if you could open a docs issue that would be great. It would be good to start this in the next week!
